Credentials with access token (oauth)

kingosticks opened the pull request:
I don't know if this oauth stuff really belongs in core, it doesn't feel quite right there so I added a new module. That new module could be useful standalone so it makes sense. If someone wants to take this and do something else that is fine by me. 

This also leaves the token stuff a bit messy. We now provide two ways to get an access token:
1. `session.token_provider().get_token("your,scopes")` using keymaster (Mercury)
2. `session.spclient().auth_token()` using login5 (HTTP) 

Both methods work (for session auth and playback) when you authenticate your session using a password or stored credentials. However, method 1 doesn't work when you authenticate your session using a spotify token (obtained using either method). 

I think we want to get rid of this annoying pitfall. We could:
a) Get rid of method 1 altogether
b) Method 1 use method 2 under the hood
c) Change session authentication so when stored-creds are not used, it auths to obtain them and then re-auths using them.

Fixes #1308

kingosticks commented:
I pulled out the login5 stuff from here since it's not actually required to implement the oauth login flow. But I did leave in the `session.auth_data()` hook required to make that work at a later date. Should make this easier to merge.

kingosticks commented:
If it mentions password then that's not using --token mode. Sorry about the bad compile, I was trying to improve it last night and then GitHub went down leaving it in a mess. I'll sort that out later hopefully

MarvAmBass commented:
anyway thanks for your work! hope we get spotify working soon :)

kingosticks commented:
Oh, and yes sorry, you meant the actual example. Yes, that still needs updating. I was only using that for testing (before they deprecated password).

MarvAmBass commented:
As I understood your code, the `examples/get_token.rs` is not yet using your oauth module. no wonder it didn't work.

regarding the OAuth way, does every user need to register a client?! or do we fake/emulate the spotify app and force spotify to redirect to localhost?

kingosticks commented:
We can keep using Spotify's desktop client ID and either pop in our own redirect Uri and do it like them, or not bother and just have the redirect fail (harmless) but then the user has to manually provide the Auth code back to our code somehow. If you run librespot in this PR you can see both modes:
```
cargo run --no-default-features -- --cache . --token ""
```
And
```
cargo run --no-default-features -- --cache . --token "" --token-port 0
```

MarvAmBass commented:
ohhh nice thanks for the hint it seems that 

```
cargo run --no-default-features -- --cache . --token ""
```

should work!

dbalague commented:
Tried it on a Raspberry Pi 5 with cargo 1.80.1 and it worked like a charm!

marcelveldt commented:
Just quickly tried out the changes in this PR but it doesn't work when using a token that was created using my own PKCE flow. So imagine you have your own (app specific) PKCE oauth flow to get the access token and pass that to librespot. It will fail on what I believe is a missing scope.


`
[2024-08-16T09:47:39Z INFO  librespot_core::session] Connecting to AP "ap-gew4.spotify.com:4070"
[2024-08-16T09:47:39Z INFO  librespot_core::session] Authenticated as '#######' !
[2024-08-16T09:47:39Z INFO  librespot_core::session] Country: "##"
[2024-08-16T09:47:39Z INFO  librespot_core::spclient] Resolved "gew4-spclient.spotify.com:443" as spclient access point
[2024-08-16T09:47:39Z ERROR librespot_core::mercury] error 403 for uri hm://keymaster/token/authenticated?scope=playlist-read&client_id=65b708073fc0480ea92a077233ca87bd&device_id=ce8d71004f9597141d4b5940bd1bb2dc52a35dae
[2024-08-16T09:47:39Z ERROR librespot_playback::player] Unable to load audio item: Error { kind: Unavailable, error: Response(MercuryResponse { uri: "hm://keymaster/token/authenticated?scope=playlist-read&client_id=65b708073fc0480ea92a077233ca87bd&device_id=ce8d71004f9597141d4b5940bd1bb2dc52a35dae", status_code: 403, payload: [[123, 34, 99, 111, 100, 101, 34, 58, 52, 44, 34, 101, 114, 114, 111, 114, 68, 101, 115, 99, 114, 105, 112, 116, 105, 111, 110, 34, 58, 34, 73, 110, 118, 97, 108, 105, 100, 32, 114, 101, 113, 117, 101, 115, 116, 34, 125]] }) }
`

kingosticks commented:
Yes, I tried to explain this in the first message. I think we can sort that out in another PR.

schickling commented:
Hopefully not too out of context for the scope of this PR but if I understood @roderickvd correctly, then moving to the HTTP-based auth mechanism would also enable using librespot in the browser via WASM? (Any additional context on this would be appreciated)

zaciam commented:
I have tested this pull request. I found an issue and a bug. There is an http server involved used to get the token code during the authentication flow with --token parameter + empty string.

1.    Issue: The port for this http server is fixed to 5588 and should be configurable.
2.    Bug: Librespot crashed when I open `http://127.0.0.1:5588/`.

mseravalli commented:
tried on a remote server to which I connected with `ssh -L 5588:127.0.0.1:5588 [...]` and the PR fails in at the unwrap: https://github.com/librespot-org/librespot/pull/1309/files#diff-a190adb0e5dac47eecedd07f77496b9d61d81b446f314411e75730044e0f4238R22 is that a supported use case or should everything run on the local network where there's already an authenticated running Spotify application?

roderickvd commented:
> Hopefully not too out of context for the scope of this PR but if I understood @roderickvd correctly, then moving to the HTTP-based auth mechanism would also enable using librespot in the browser via WASM? (Any additional context on this would be appreciated)

I don't think so, not until someone comes up with an original (i.e. non-decompiled) version of the `playplay` HTTP endpoint that gets the track decryption keys. Without that, we still require TCP-based Mercury to get the keys, and that won't work in WASM right? No WASM expert.

> I have tested this pull request and found a little bug. There is an http server involved used to get the token code during the authentication flow with `--token ""`. Librespot crashes when `http://127.0.0.1:5588/` is opened without the redirect code.

Anyone who can confirm and/or fix this?

Back from holidays, starting my review now.

roderickvd commented on core/src/session.rs:
```diff
@@ -77,6 +77,7 @@ struct SessionData {
     client_brand_name: String,
     client_model_name: String,
     connection_id: String,
+    auth_data: Vec<u8>,
```
If (and only if) more ergonomical, you could consider using `Bytes` instead of `Vec<u8>`. I think we pull in that dependency anyway, and it may make lifetimes & ownership around this networking I/O easier.

kingosticks replied:
I have no experience using `Bytes`. But given someone might want to use this crate without the rest of librespot, maybe it's not worth forcing that extra dep on them?

HirbodBehnam replied:
Is it actually a good idea to use `Bytes` here? Based on the code usage, the only usage of `auth_data()` method is in `get_token.rs` in examples which is used as `Credentials::auth_data` which is an owned `Vec<u8>`. If we were using `Bytes`, we had to create an owned `Vec<u8>` for it which I think defeats the purpose of the `Bytes` zero-copy clone method because of `Vec<u8>` allocation.
However, something else we can do is to change the `Credentials::auth_data` to be `Bytes` instead of `Vec<u8>`. I'm not sure what implications this change would have but we can investigate.
Also, if you are committed to use `Bytes` here, you can use these two functions:
```rust
    pub fn auth_data(&self) -> Bytes {
        self.0.data.read().auth_data.clone() // this is zero copy!
    }

    pub fn set_auth_data(&self, auth_data: &[u8]) {
        self.0.data.write().auth_data = Bytes::copy_from_slice(auth_data);
    }
```
Each `Bytes` can be converted to an `Vec<u8>` using [`to_vec`](https://docs.rs/bytes/1.7.1/bytes/struct.Bytes.html#method.to_vec) method. Also, based on the `Cargo.toml`, the `librespot-core` package contains the `Bytes` library as its dependency.

kingosticks replied:
Thanks for looking into this.

> Also, based on the Cargo.toml, the librespot-core package contains the Bytes library as its dependency.

The idea is that his new librespot-ouath crate can be used independently of librespot (e.g. someone's random code that uses Spotify's Web API) so the existing dependencies of librespot-core don't come into it.

roderickvd replied:
Like I said, no worries, only if it were more ergonomical.

roderickvd commented on Cargo.toml:
```diff
@@ -49,6 +49,10 @@ version = "0.5.0-dev"
 path = "protocol"
 version = "0.5.0-dev"
 
+[dependencies.librespot-oauth]
```
Good idea to extract it to a new crate. Should require an update to https://github.com/librespot-org/librespot/blob/dev/publish.sh too.

roderickvd commented on oauth/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+name = "librespot-oauth"
+version = "0.5.0-dev"
+rust-version = "1.73"
+authors = ["Paul Lietar <paul@lietar.net>"]
```
I think you should change it to your name. After all, you wrote this part, not Paul.

roderickvd commented on oauth/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+version = "0.5.0-dev"
+rust-version = "1.73"
+authors = ["Paul Lietar <paul@lietar.net>"]
+description = "Spotify OAuth"
```
Following the other crates, for consistency maybe a bit expanded like "The OAuth logic for username and password authentication with Spotify"

roderickvd commented on oauth/src/lib.rs:
```diff
@@ -0,0 +1,152 @@
+use url::Url;
+
+fn get_code(redirect_url: &str) -> AuthorizationCode {
+    let url = Url::parse(redirect_url).unwrap();
```
Moving from 0.4 to 0.5 I spent a lot of time making the code panic-free by purging all `unwrap` and `expect`s. Could you have this and the below return `Result` too?

kingosticks replied:
OK, I worked through most of this last night. As a part-time Rust contributor, I'm guilty of unwrapping everything rather than handling errors properly so this stuff was a bit new. I probably haven't got it as nice as it could be but it's better.

roderickvd replied:
Looked through it. Unless I'm mistaken, I still see quite a lot of code that could panic due to unwrapping. How can I help you set it up to return e.g. a `Result<AuthorizationCode, core::Error>` (the other library parts should offer examples on how to but let me know if you need some pointers).

kingosticks replied:
I've worked on doing it but I've yet to push that work as I didn't quite finish. I'll try and sort it now.

kingosticks replied:
There we go

roderickvd commented:
Sorry, hit the wrong button adding all the review comments individually. Let me know if this is acceptable for you to work on before merging. Would also be great if you could add to the changelog.

kingosticks commented:
Sure, I'm currently on holiday but will do on my return next week.

SuisChan commented:
> I don't think so, not until someone comes up with an original (i.e. non-decompiled) version of the playplay HTTP endpoint that gets the track decryption keys. Without that, we still require TCP-based Mercury to get the keys, and that won't work in WASM right? No WASM expert

I'm not a wasm expert either, but I think my workaround should work in WASM (I haven't tested it WASM, but I'm pretty sure), but I'm not going to publish it until it becomes critical or spotify gets lossless support : (

G-M0N3Y-2503 commented:
I hope this isn't just noise and may be of help. 
But a while back I implemented [OAuth for the Spotify API in Rust targeting WASM](https://github.com/G-M0N3Y-2503/spotifyrs/blob/1ab80483c9af997d5e04d0007e49ff0dc0c0e17d/spotify-api/src/authorisation.rs) I never got around to doing anything with it so I'll leave it here for reference.
If you find it particularly helpful, I would appreciate attribution.

3052 commented:
> > I don't think so, not until someone comes up with an original (i.e. non-decompiled) version of the playplay HTTP endpoint that gets the track decryption keys. Without that, we still require TCP-based Mercury to get the keys, and that won't work in WASM right? No WASM expert
> 
> I'm not a wasm expert either, but I think my workaround should work in WASM (I haven't tested it WASM, but I'm pretty sure), but I'm not going to publish it until it becomes critical or spotify gets lossless support : (

I would advise anyone not to hold their breath on this. dude has been teasing this literally for years:

https://github.com/librespot-org/librespot-java/discussions/421

so @SuisChan either post the code or stop talking about it.

3052 commented:
> I'm not a wasm expert either, but I think my workaround should work in WASM (I haven't tested it WASM, but I'm pretty sure), but I'm not going to publish it until it becomes critical or spotify gets lossless support : (

I would advise anyone not to hold their breath on this. dude has been teasing this literally for years:

https://github.com/librespot-org/librespot-java/discussions/421

so @SuisChan either post the code or stop talking about it.

SuisChan commented:
> > I'm not a wasm expert either, but I think my workaround should work in WASM (I haven't tested it WASM, but I'm pretty sure), but I'm not going to publish it until it becomes critical or spotify gets lossless support : (
> 
> I would advise anyone not to hold their breath on this. dude has been teasing this literally for years:
> 
> [librespot-org/librespot-java#421](https://github.com/librespot-org/librespot-java/discussions/421)
> 
> so @SuisChan either post the code or stop talking about it.

You know, this is a bit funny. but let's make this clear. like I said, this thing is real, there's an old version (clone) on GitHub somewhere (my orig was wiped out [link](https://github.com/github/dmca/blob/master/2022/02/2022-02-25-spotify.md)).

And like I said, I won't share it unless it's really necessary (completely broken/disconnected tcp connection or something like that), I need a reason to share the decryptor, ok? I won't repeat it again. Period.

You can say whatever you want, it won't change anything. 🤷

3052 commented:
> And like I said, I won't share it unless it's really necessary (completely broken/disconnected tcp connection or something like that), I need a reason to share the decryptor, ok? I won't repeat it again. Period.

you are talking about your mystery code for years, and not releasing it. this is not only pointless, but it actively hurts the community, because some people will just sit around waiting for you to release it, rather than making something themself. so stop it.

roderickvd commented:
Let's keep it constructive. I've seen the code before the link was deleted because Spotify told us to. And they could because it was a decompiled version, which is not allowed. Best we can do is come up with an original implementation that's not their IP.

SuisChan commented:
> people will just sit around waiting for you to release it, rather than making something themself

1. Code was published before, believe it or not. 
2. Yes, because I spent my time on it, so it's up to me to decide when to publish, no?

3052 commented:
> Let's keep it constructive. I've seen the code before the link was deleted because Spotify told us to. And they could because it was a decompiled version, which is not allowed. Best we can do is come up with an original implementation that's not their IP.

you wanna be constructive, then post the decompiled code. if GitHub dont allow it, then post somewhere else and link to it here. I would argue that talking about mystery code is highly unconstructive, so lets make it concrete. if the decompiled code is made widely available, then people can use it to come up with original implementations.

kingosticks commented:
I think ideally I'd move this out of `get_setup()` and into `main()`. I think it makes more sense to start a server there and it'd simplify the diff a bit.

photovoltex commented:
@kingosticks, you mention in the `get_token` example that the initial session created with the token wouldn't work with keymaster. Wouldn't it make more sense to add the "workaround" or solution to that problem into the `connect` method of `Session`?

Due to handling that case only in the `get_token` example, starting librespot with a token will not work unless you cache the session and restart librespot with the cached session.

kingosticks commented:
@photovoltex yep you are totally right. When I first made this PR I also implemented the login5 side of things, and replaced the token master calls. I then removed all that in an attempt to make this PR smaller and simpler, which has exposed that issue again.

SilverMira commented:
By chance, is it normal for playback to not work when using my own client_id? I'm requesting the same scopes but for whatever reason I'm only getting 403 forbidden calling `AudioItem::get_file()` when using my own client_id, but the same works totally fine while using the official client's ID.

kingosticks commented:
I think it's buried in the comments here but essentially, if you're using our default list of scopes then that is normal. The default scopes we are requesting here are the same as what the desktop client requests. It seems at least one of the scopes isn't allowed for client IDs other than Spotify's. If you trim it down to just "streaming" you should be able to use your own client ID. Ideally we'd work out what's the minimum subset of universally allowed scopes and what's needed to make everything in librespot work, but that's outside my interest.


I hate GitHub's hiding comments "feature".

SilverMira commented:
> If you trim it down to just "streaming" you should be able to use your own client ID, I think I tested that... maybe you can reconfirm? 

Turns out setting `SessionConfig::client_id` to my own client_id was what resulted in the Forbidden 403 when calling `AudioItem::get_file`, I thought at first `SessionConfig::client_id` must be the same client_id as what was used in the oauth flow? Still, no idea whether this behavior is expected or not.

Now what works for me:
1. OAuth flow with only scope "streaming" using my own client_id
2. Get credentials with `Credentials::with_access_token()`
3. Create and use `session2` as the playback session, noting that `SessionConfig::client_id` must use the official's ID, following the example https://github.com/librespot-org/librespot/blob/5093a88e5f7162d2a453c005c60530abaf9bcc47/examples/get_token.rs#L22-L50
4. `AudioItem::get_file` and friends works, able to stream audio data

kingosticks commented:
Thanks for clarifying. Yes, that is expected. Maybe I didn't explain well before but I can see how the example is confusing. Essentially, you can use any client ID you want in step 1, it's a regular Spotify OAuth flow as per their public developer documentation. But you must use one of Spotify's client IDs when creating a librespot `Session`. Your client ID doesn't have the required permissions to access their internal APIs. I will try and clarify the example and the OAuth module docs.

kingosticks commented:
This is in a better state now. Still zero tests (sorry, World) but `unwrap`-free and slightly better documentation. 

One bit I still don't like is my `OAuthToken` type.  I didn't want to return the oauth crate token since it's overly complex to use and I'd rather own the API interface types. I didn't want to pull in our existing token type from `core` because that adds an otherwise unnecessary dependency. Maybe I should provide a `from<OAuthToken>` implementation in `core`. 

Arguably all the newer token stuff in core could be pulled out into a useful standalone crate but I don't quite see the point (yet).

roderickvd commented on oauth/src/lib.rs:
```diff
@@ -0,0 +1,227 @@
+    #[error("Invalid Redirect URI {uri} ({e})")]
+    InvalidRedirectUri { uri: String, e: url::ParseError },
+
+    #[error("Failed to recieve code")]
```
typo: recieve -> receive

roderickvd commented on oauth/src/lib.rs:
```diff
@@ -0,0 +1,227 @@
+) -> Result<OAuthToken, OAuthError> {
+    // Must use host 127.0.0.1 with Spotify Desktop client ID.
+    let redirect_address = SocketAddr::new(
+        IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
```
Mindfart: how does that work for IPv6-only hosts? If Spotify supports that at all?

kingosticks replied:
I could certainly try setting the redirect host ip to ::1 and test if they accept that, my random guess is they won't, buy it's quick enough to try. As it is, someone running ipv6-only would have to use the stdin version of the code recovery as  a workaround (`--oauth-port 0`). I don't have an ipv6 only system to test with so I'm not that keen on trying to implement it.

roderickvd replied:
I agree, just a mindfart triggered by #1307. Don't worry about it.

kingosticks replied:
Using ipv6 localhost in the `redirect_uri` param for the Oauth flow did actually work so I think it would be possible to support this properly... but I'd still rather leave it for someone who could test it properly.

kingosticks replied:
@roderickvd I fixed the last todo so now takes a `redirect_uri` argument instead of just the port. If you set that to http://[::1]/foo etc, you'll get an IPv6 socket.

roderickvd commented on oauth/src/lib.rs:
```diff
@@ -0,0 +1,227 @@
+    };
+    Ok(OAuthToken {
+        access_token: token.access_token().secret().to_string(),
+        refresh_token: token.refresh_token().unwrap().secret().to_string(),
```
One more unwrap to go!

roderickvd commented on oauth/src/lib.rs:
```diff
@@ -0,0 +1,227 @@
+use url::Url;
+
+#[derive(Debug, Error)]
+pub enum OAuthError {
```
Would you also provide an `impl From<SpircError> for Error`?

kingosticks replied:
Sorry, not sure I follow. How/why would this be used here?

roderickvd replied:
Sorry, I meant `impl From<OAuthError> for core::Error` like so:

https://github.com/librespot-org/librespot/blob/f6473319f6309b2702a20ff5371763c48e129676/connect/src/spirc.rs#L48

kingosticks replied:
The only thing not ideal with that, is having to pull in librespot `core` to do it..

roderickvd replied:
Possibly the other way around: put the implementation in core::error so that pulls in oauth.

roderickvd reviewed (changes requested):
I found two more unwraps, if you could fix those?

roderickvd reviewed (approved):
Even with tests! Perfect 😁

roderickvd commented:
Perfect. I am going to merge this now. I don't have time for a release this weekend, so let's target next week for that. That leaves room coming days to give this a good shake-down, and putting in any last-minute PRs.

There are a couple of almost-finished good PRs in there, would be nice if they could get polished in time for release.

Losses commented:
Wow, is there a detailed document about how to use this  for now?

Losses commented:
Okay, a quick note for users who is feeling confusing about this:

## OAuth Authentication

Password login has been deprecated and is no longer supported. Instead, OAuth authentication must be used. Follow the steps below to authenticate using OAuth:

1. Run the following command to initiate the OAuth process:

    ```sh
    librespot --cache YOUR_CONFIGURATION_PATH -j
    ```

2. If there is no token in the specified directory, you will see a log similar to this:

    ```
    [INFO  librespot] librespot 0.5.0-dev b68516c (Built on 1980-01-01, Build ID: 315532800, Profile: debug)
    Browse to: https://accounts.spotify.com/authorize?response_type=code&client_id=...
    ```

    **Note:** This URL contains sensitive information. Please be cautious and avoid sharing it publicly.

3. Open the provided URL in a web browser to complete the login process.

4. If your device is remote, you have two options:
    - Complete the login process locally and then copy the token directory to the remote location.
    - Use port forwarding to complete the login process remotely, you can see the port being forwarded from the log:

```
[INFO  librespot_oauth] OAuth server listening on 127.0.0.1:5588
```

Good luck.

kingosticks commented:
I'll update the wiki and any other documentation I can find regarding auth. I might leave some for posterity (clearly marked as deprecated).

Regarding point 2 above, i don't believe there's anything fundamentally sensitive in that URL, particularly as we're using Spotify's client ID which is easy to discover and use. But you raise a great point and there is sensitive info in the trace messages that follow as it displays the secret token values. When I was developing this I found that detail invaluable, but now there's an example in the oauth module which can provide the same detail, so we could redact the logging here.

For point 4, there's an alternate mode for this situation where instead of spawning a server to accept the response URI, librespot instead prompts you to manually enter that response URI. This means you can complete the oauth dance on your local machine and when the redirect fails (because nothing is listening on your local machine), just copy and paste the URL into librespot's prompt. There's a few ways to activate this mode, the easiest is to use `--oauth-port` option to *NOT* specify a port. I wanted to avoid adding yet another option to explicitly control this mode but I'm not sure where I ended up is the nicest, even once it's documented.

spockfish commented:
Hi @kingosticks,

First of all: thanks for your work! I think that everyone running Librespot was waiting for this to fix the authentication stuff.

With respect to 4: would it also be an idea to be able to provide the redirect url manually? I'm aware that the user then has the responsibility to provide the correct one of course, but for the 'remote' situation this might be nice...

Anyways, again thanks for fixing the authentication!

spockfish commented:
Question: everything is working, except that I can't see/find where the token is stored for the next time. So right now I need to repeat the procedure whenever I start librespot.

kingosticks commented:
With the exception of the port, there isn't actually a choice when it comes to what redirect Uri to use, it has to match what's registered for the particular client ID being used. I think `--oauth-port 0` for headless/remote users should be adequate.

The token isn't stored. The token is exchanged for Spotify stored credentials, which are then saved in the cache folder as usual. You should run with `--cache <some-folder>`. I'm not sure why caching isn't enabled by default.

kingosticks commented:
@roderickvd what do you think about making the the system-cache enabled by default in 0.5? Having to do the oauth flow every time is bonkers and it shouldn't be the default. Personally I'd go the whole hog and have all caching enabled by default.. anyone who cares should be able to handle the change. And those that don't care can enjoy better performance.

Losses commented:
@kingosticks What about print a large warning when cache directory is not available to make users aware of what are they doing?

roderickvd commented:
Makes sense to me, to enable it by default. Would you put in a change?

spockfish commented:
> I think --oauth-port 0 for headless/remote users should be adequate.

Challenge for me is that I provide Librespot as part of a appliance (RoPieee) that does not require any kind of (linux) knowledge to get things rolling.

And right now I don't have a clue how to create an authorization feature that's still relatively easy to do for non-geeks like us...

I tried this by creating a small reverse proxy and change the port number of the redirect uri... but that fails. The redirection and all works, but I get an error message from Librespot.

Maybe I'm overlooking something here (or I'm misunderstanding things), but I appreciate your thoughts on this...

kingosticks commented:
Ideally your users would be able to copy and paste the failing redirect Uri into whatever admin console you presumably already have. You can then start librespot with `--token`. I don't believe there is any way to highjack the redirect, that's no accident.

Otherwise, reimplement the oauth handler on your own domain with your own client ID,  request streaming scope, and display the resulting access token for the user to then paste into your admin console as above. A lot more hassle for a less ugly flow, but still essentially copying something from one window to another. This also requires you to get your own Spotify client ID and last time I checked, Spotify were hopeless at approving these.

Or maybe your users can auth using Spotify connect? Ideally we'd be able to provide the device flow but I don't know if anyone reversing this https://community.spotify.com/t5/Spotify-for-Developers/Device-Authorization-Grant-authentication-flow-for-custom/td-p/5485468

SilverMira commented:
Just an idea: I was sort of able to hijack the redirect by spawning a WebView of my control to the oauth url and listening for navigation to the callback url, this works even if callback urls are custom protocol like spotify-auth:// on android.

This depends if you have your own "frontend" configuration app though

kingosticks commented:
That works because you are listening on the same machine where the redirect is happening. In the headless case you simply cannot do that.

kingosticks commented:
> Personally I'd go the whole hog and have all caching enabled by default.. anyone who cares should be able to handle the change. And those that don't care can enjoy better performance. But I don't want to have a big discussion on the data cache here, I'll make a new issue if there's interest in that part.

I've been thinking more about this and I'm less opinionated now, to the point I'm not going to bother. To have it default enabled you need a default location. Could be the cwd or could be the XDG dir. Then you need the (existing) option to specify somewhere else and a new option to disable it entirely. And that's got to play nice with the general cache directory too. This is all more work and it's not interesting work to me, I've enough boring work to do as it is. I'll add a warning as suggested, and be done with it.

TDethlefs commented:
## OAuth Authentication
I used a third Option with Curl:

1. Run the following command on the remote device to initiate the OAuth process:
   ```shell
   librespot --cache YOUR_CONFIGURATION_PATH -j
   ```
2. Open the provided URL in a web browser to complete the login process until you are redirected to a "127.0.0.1:5588..." URL that cannot be opened. Copy that URL, SSH-Login to your remote Device and run curl with that URL:
  ```shell
  curl "http://127.0.0.1:5588/login?code=AQC..."
  ```
 The double quotes around the URL after "curl" are important as the URL includes a "&" that is interpreted by the shell if not in double quotes. 
 
 Good Luck.

kingosticks commented:
Which is like using `--oauth-port 0` and pasting in your redirect Uri when prompted, but adds using curl also. I'm not convinced that adding more software to the process helps but whatever works for you.

armanhi commented:
Quick question: how is the user session refreshed/kept on librespot? I can see the new OAuth "get_access_token" method returns a refresh_token, but it doesn't seem to be used anywhere in the application.

armanhi commented:
Thank you for your reply!

So, if I understand it correctly, the access token lasts indefinitely? I've always assumed it should be refreshed. Also, in the Spotify docs (https://developer.spotify.com/documentation/web-api/tutorials/code-flow) is says the "https://accounts.spotify.com/api/token" response contains a "expires_in" field, meaning "The time period (in seconds) for which the access token is valid.". That's where my question comes from.

I'm asking mainly because of something that happened to me the other day. I've generated a credentials.json file using zeroconf from librespot-auth (https://github.com/dspearson/librespot-auth), copied it to another machine and, using spotify-player, after a while, probably 1 hour, I started getting the "Bad credentials" error, so I thought the reason was the 3600s duration returned then the access token is obtained. Maybe the zeroconf expires?

Just one more question: does the credentials.json file support the "https://accounts.spotify.com/api/token" response I mentioned above (containing the access_token directly) or just the zeroconf json contract?

kingosticks commented:
No, all Spotify access tokens expire. The token you get back literally has an expiry.

Once we login using a token, we can request  a credential blob aka reusable credentials aka stored credentials. This blob does not expire and can be used instead of an access token next time. 

I don't really understand the other part of your question but I believe the Zeroconf auth blob is something different:

> The encrypted blob received in this ZeroConf message must not be saved. On a successful login, the SpCallbackConnectionNewCredentials() callback will be invoked with the credentials to save to persistent storage.

https://developer.spotify.com/documentation/commercial-hardware/implementation/guides/zeroconf

I don't remember much about the Zeroconf stuff, I don't use it. If you have questions or problems with it, start a new discussion post so others can see it and help newer.

kingosticks commented:
But I should also add, I don't know exactly what librespot-auth does. It's always been possible to produce a credentials.json file via zeroconf using regular librespot. And I don't know what 'spotify-player' is.

armanhi commented:
I mean, they do use librespot under the hood, but since Spotify's change I definitely understand it's kinda tough to keep everything working as usual.

Is it possible to pass a valid access token as argument to librespot? Should the credentials.json file work if I have a json with the format
`{
"access_token":"ACCESS_TOKEN",
"token_type":"Bearer",
"expires_in":3600,
"refresh_token":"REFRESH_TOKEN",
"scope":"SCOPES_LIST"}
` ?

kingosticks commented:
Yes, use --help to see the argument name.

No, that won't work. Credentials.json should be created for you by librespot.

armanhi commented:
Awesome, thank you so much!

kingosticks commented:
I (finally) added some missing option wiki documentation and clarification. I will try to add something to the authentication wiki page at some point.

PureTryOut commented:
@kingosticks I'm probably just incredibly dumb, but using librespot-oauth 0.6.0 I still get `Error { kind: PermissionDenied, error: LoginFailed(BadCredentials) }` currently.

With the following code:

```rust
let token: OAuthToken = get_access_token("...", REDIRECT_URI, vec![...]).unwrap();
let session_config = SessionConfig::default();
let session = Session::new(session_config, None);

let credentials = Credentials::with_access_token(token.access_token);
let result = session.connect(credentials, false).await.unwrap();
```

Is this expected?
